src/lib/snapp.ts

```diff
@@ -268,6 +272,83 @@ function isProof(typ: any) {
   return false; // TODO
 }
 
+type Statement = { transaction: Field; atParty: Field };
+
+function toStatement(
+  self: Party,
+  tail: Field,
+  state: ProtocolStatePredicate,
+  checked = true
+) {
+  // TODO hash together party with tail in the right way
+  if (checked) {
+    let atParty = Ledger.hashPartyChecked(toParty(self));
+    let protocolStateHash = Ledger.hashProtocolStateChecked(
+      toProtocolState(state)
+    );
+    let transaction = Ledger.hashTransactionChecked(atParty, protocolStateHash);
+    return { transaction, atParty };
+  } else {
+    let atParty = Ledger.hashParty(toParty(self));
+    let protocolStateHash = Ledger.hashProtocolState(toProtocolState(state));
+    let transaction = Ledger.hashTransaction(atParty, protocolStateHash);
+    return { transaction, atParty };
+  }
+}
+
+function checkStatement(
+  { transaction, atParty }: Statement,
+  self: Party,
+  tail: Field,
+  protocolState: ProtocolStatePredicate
+) {
+  // ATM, we always compute the statement in checked mode to make assertEqual pass
+  // TODO: are checked / unchecked versions supposed to be compatible?
+  let otherStatement = toStatement(self, tail, protocolState, true);
+  atParty.assertEquals(otherStatement.atParty);
+  transaction.assertEquals(otherStatement.transaction);
+}
+
+let mainContext = undefined as
+  | {
+      witnesses?: unknown[];
+      self: Party & { predicate: AccountPredicate };
+      protocolState: ProtocolStatePredicate;
+    }
+  | undefined;
+
+function withContext(context: typeof mainContext, f: Function) {
```

Comment:
Can we change this to not rely on a global variable? This could lead to scary bugs later.

We could pass it into `f` or flex our inner JavaScript and `bind` the context at the callsite to this: `f.apply(context)` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply

Reply:
I'm really glad you brought this up. When I added this originally, I guess it was fine because both compiling and proving were synchronous, so there was nothing that could conflict with the global variable while they were running (JS is single-threaded). However, since then, the Pickles provers became async. So there are probably real bugs in the code right now if you'd run `prove` multiple times concurrently. I'll have to fix this!
It's not as simple as passing it into `f` though. The reason I made it a global variable is that the pickles circuit needs to access private inputs, but it has no arguments to do so. The interface of Pickles sort of forced me to use a global variable, to store context that the circuit could access. The `f.apply(context)` thing is creative, hadn't thought of that, maybe I can even use that to solve the problem.

Reply:
I couldn't come up with a solution with a behaviour that's safer than global variables.
So the way I fixed the potential bug for now, is observing how many times the mainContext is accessed, and throw an error if it's more than the expected number of times. This should reliably detect concurrent accesses. It's not ideal, but at least it protects correctness: https://github.com/o1-labs/snarkyjs/pull/123/commits/84a014996eaadf157dd7fc890b5e36fc51993f74

Reply:
Good workaround!

> that the pickles circuit needs to access private inputs, but it has no arguments to do so
Is this something we can fix by changing the pickles circuit interface? Can you make in issue tracking this? (in Mina repo?)

Reply:
That change to pickles seems very far-reaching... not sure if it should be changed. But made an issue to surface the problem: https://github.com/MinaProtocol/mina/issues/10815